File: packages/ingestion/priority/urlnorm_test.go
package priority

import (
	"encoding/json"
	"os"
	"testing"
)

func TestNormalizePageURL(t *testing.T) {
	raw, err := os.ReadFile("../../../test-fixtures/url-normalization/vectors.json")
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var fixture struct {
		Vectors []struct {
			Name string `json:"name"`
			In   string `json:"in"`
			Want string `json:"want"`
		} `json:"vectors"`
	}
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatalf("parse fixture: %v", err)
	}
	if len(fixture.Vectors) < 50 {
		t.Fatalf("fixture suspiciously small: %d rows", len(fixture.Vectors))
	}
	seen := make(map[string]bool, len(fixture.Vectors))
	for i, tc := range fixture.Vectors {
		if tc.Name == "" {
			t.Fatalf("row %d: empty name (schema guard — a malformed row must not silently pass)", i)
		}
		if seen[tc.Name] {
			t.Errorf("duplicate fixture name %q", tc.Name)
		}
		seen[tc.Name] = true
		if tc.In == "" && tc.Name != "empty input stays empty" {
			t.Fatalf("row %q: empty input on a row not named for it", tc.Name)
		}
		if got := NormalizePageURL(tc.In); got != tc.Want {
			t.Errorf("%s: NormalizePageURL(%q) = %q, want %q", tc.Name, tc.In, got, tc.Want)
		}
	}

	// route_map's btree primary key refuses an oversized pattern, and one such
	// row aborts the transaction that writes every other route for the project.
	long := "https://app.example.com"
	for i := 0; i < 400; i++ {
		long += "/abcdefgh"
	}
	if got := NormalizePageURL(long); got != "/too-long" {
		t.Errorf("oversized path: NormalizePageURL() = %q, want /too-long", got)
	}
}

// The two kinds reach NormalizePageURL in different shapes: error groups from
// the raw browser URL, friction groups from a value fingerprint.ts already
// normalized (origin kept, _ctx_ segments dropped). Both must land on one
// pattern or route_map weights only half the incidents on a page.
func TestConvergesAcrossBothStampingPaths(t *testing.T) {
	cases := []struct{ name, raw, preNormalized, want string }{
		{
			"embedded panel with a context segment",
			"https://x.cdn.prod.example.net/a/b/issue-context/_ctx_abc/",
			"https://x.cdn.prod.example.net/a/b/issue-context",
			"/a/b/issue-context",
		},
		{
			"ordinary page",
			"https://app.test/orders/4021",
			"https://app.test/orders/:id",
			"/orders/:id",
		},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			gotError := NormalizePageURL(c.raw)
			gotFriction := NormalizePageURL(c.preNormalized)
			if gotError != c.want || gotFriction != c.want {
				t.Errorf("error path = %q, friction path = %q, want both %q", gotError, gotFriction, c.want)
			}
		})
	}
}

File: packages/worker/src/friction/__tests__/urlnorm.test.ts
import { readFile } from 'node:fs/promises';
import { describe, expect, it } from 'vitest';
import { canonicalPattern, normalizePageUrl } from '../urlnorm.js';

interface Vector {
  name: string;
  in: string;
  want: string;
  quirk?: boolean;
}

const fixture = JSON.parse(
  await readFile(
    new URL('../../../../../test-fixtures/url-normalization/vectors.json', import.meta.url),
    'utf8',
  ),
) as { vectors: Vector[] };

describe('normalizePageUrl (cross-language contract)', () => {
  it('fixture is present and substantial', () => {
    expect(fixture.vectors.length).toBeGreaterThanOrEqual(50);
  });

  it.each(fixture.vectors)('$name', ({ in: input, want }) => {
    expect(normalizePageUrl(input)).toBe(want);
  });
});

describe('canonicalPattern', () => {
  it('strips an origin prefix', () => {
    expect(canonicalPattern('https://app.example.com/assets/:id')).toBe('/assets/:id');
  });

  it('origin-only becomes root', () => {
    expect(canonicalPattern('https://app.example.com')).toBe('/');
  });

  it('path-only passes through', () => {
    expect(canonicalPattern('/assets/:id')).toBe('/assets/:id');
  });
});
